### ISSUE TITLE
Long-Term Possibility of Flatpak Compatibility?

### ISSUE THREAD

**Issue by Eoin-ONeill-Yokai** [robbert-vdh/yabridge#135]
### Feature description

There's been some discussion in other issues regarding users trying to use Flatpak versions of professional DAWs with Yabridge -- specifically #114 and #111 where it's been clarified that it's not technically possible to use flatpak w/ yabridge due to sandboxing at the moment.  This has been mentioned in bitwig/bitwig-studio-flatpak#24 as well, which might be worth following up or discussing. Maybe if bitwig and yabridge could establish some communication here, it would be helpful!

My understanding of flatpak is pretty limited, but I was wondering if there could be some way to support flatpak distributions of DAWs while also supporting yabridged vst instruments.

## Why?
While most DAWs provide some kind of `.deb` installer, there are certain systems (Fedora, Arch) where installing the flatpak version could lead to better stability (due to correct library usage within the DAW) while providing an easier installation process where software versions can be specified and maintained easily via Flathub.  Because of this, it would be nice to eventually have Yabridge be compatible with flatpak daws.

## Flatpak Extensions
My limited research on Flatpak resulted in finding out about `Flatpak Extensions`, which I think would allow Yabridge to provide a "add-on" environment onto standard Bitwig installations via flatpak. My understanding would be that it would look something like this:

```
flatpak install org.bitwig.Bitwig.Yabridge
```

Where the last domain represents the desired extension on top of the `org.bitwig.Bitwig` environment. This would provide the necessary environment for Yabridge (my understanding is basically LibBoost + Wine-Staging, but there's perhaps more necessary) which would then be optionally included within the sandboxed Bitwig instance. This extension would need to be uploaded and maintained within the flathub repository.

There might be more I'm missing here, as my understanding of this topic is pretty limited. If anyone happens to know more about Flatpak, it would be nice for some input here. Additionally, I'm not sure if a flatpak exension needs to be made by Bitwig, for instance, to get this working as expected or if the extension can be maintained separately by yabridge contributors. 

Anyway, I'm mostly interested in this as a user who would like to use flatpak for most of my user-side installations for stability reasons. If anyone here has any insight on the issue, it would be nice to hear it.

**Comment by robbert-vdh** [comment]
You would need to package both Wine and yabridge in such a way that they can run inside of the Flatpak runtime. One crucial issue with using Flatpak for these things remains that it was always meant to sandbox individual applications, while DAW usage often involves integrating many separate components such as plugins (which include things like yabridge and hardware-specific plugins that depend on other applications running on the same system) and connected hardware. So I'm still not convinced that distributing DAWs through Flatpak is the correct solution to the packaging problem, but that's a different matter entirely of course. :smile:

But yeah, like I said, if both yabridge and a suitable version of Wine are available within the Flatpak runtime then everything would just work. The issue then becomes that yabridge consists of two sets of components: the plugin libraries (`yabridge-{vst2,vst3}.so`), and the plugin host binaries (`yabridge-{host,group}{,-32}.exe{,.so}`), and those versions need to match. So to be able to actually use yabridge for a plugin within a Flatpak container, you'd need to either:

- Package Wine and yabridge for Flatpak, install yabridge as normal outside of Flatpak (which I'll just call the host system from now on), set up yabridge for the plugins as normal on the host system using `yabridgectl sync`, and then as the user make sure the Flatpak environment is using the same version of yabridge you installed on the host system to avoid conflicts. This would still require the Flatpak container to have access to the Wine prefixes (I think Bitwig's Flatpak already removes the sandboxing for the entire home directory, but normally those things are sandboxed).
- Do the same thing, but use the yabridge plugin libraries packaged in the Flatpak runtime instead of the ones installed on the host system and also run yabridgectl from within the container. That would reverse the dependency issue, since you now cannot use those plugins with other non-Flatpak DAWs unless you have the same version of yabridge installed outside of Flatpak.
- Install Wine inside of the sandbox, and allow the container to directly use the yabridge version installed to the host system.  This is probably the least headache introducing option since it sidesteps the whole dependency issue I talked about earlier, but this still would not work if yabridge is installed through the package manager (unless we're specifically talking about the `yabridge-bin` AUR package). In that case yabridge would be dynamically linked against Boost.Filesystem, so the same version of Boost that's used on the host system would also need to be available in the Flatpak container. That was the issue @drewofdoom had in https://github.com/bitwig/bitwig-studio-flatpak/issues/24. So this could only really work with a binary release of yabridge manually installed to `~/.local/share/yabridge`.

**Comment by Thesola10** [comment]
I believe the [Bottles project](https://github.com/usebottles/bottles) has already packaged Wine for Flatpak, and might be a plausible approach to sharing Wineprefixes between host and flatpaks (ideally).

One could imagine syncing the Wine version and runtime using a document portal between Bottles and any app using the yabridge Flatpak plugin. UX would be something like "Bitwig needs to run Windows software. Create or select a Bottle to use with Bitwig"

(ofc that approach requires work on the Bottles side of things, and potentially on yabridge too, but just imagine a standard, centralized way to manage Wineprefixes throughout the Flatpak ecosystem)

**Comment by David-Else** [comment]
Now that stable Wine itself is a Flatpak `org.winehq.Wine` can we use that? It perfect as every distro can use, but there is no instructions or talk about it here.

**Comment by robbert-vdh** [comment]
_Right now_ (this can of course change at any time) `org.winehq.Wine` and `com.bitwig.BitwigStudio` use the same Flatpak runtime so they'd in theory be able to be used together. Yabridge 4.0 at least sort of solves the out of sync issue with Flatpak+yabridge, but you'd still need yabridge packaged for Flatpak (if you expose the entire home directory than a manual/Ubuntu-style installation should also work if you run yabridgiectl outside of the Flatpak runtime, but this is also a bit hacky). So in theory you could be able to make it work, but Flatpak is simply not meant to be used with ecosystems like this so you may run into other problems.

**Comment by dreamcat4** [comment]
ok @robbert-vdh ... so in your opinion would there be any value in making a *downstream* versioned flatpak builds of yarbridge specifically targeting for bitwig only. To ensure that the wine versions can be kept in sync and lined up to match each other?

Perhaps the flatpak fully qualified name might be called `com.github.<username>/yarbridge-bitwig` or `yarbridge-for-bitwig`. Or something like that to appear as it's own seperate flatpak allongside the seperate bitwig flatpak. Since presumably we cannot make yarbridge a hard requirement of the official bitwig flatpak, at least not unless the Bitwig company ever chooses to do that.

Ok... so does that all sound like madness? or some kind of a path forwards here? (We would also need somebody else to come alone who is willing to create and maintain such a flatpak)

**Comment by robbert-vdh** [comment]
You could do that. But honestly, in my opinion Flatpak in its current form simply cannot work for software like DAWs and plugins that work together as an entire ecosystem. You may run into other issues, and once the Flatpak runtimes are starting to get upgraded you'll need to somehow upgrade the Flatpak build files for all packages at once or things may not be able to work together anymore.

**Comment by Eoin-ONeill-Yokai** [comment]
> You may run into other issues, and once the Flatpak runtimes are starting to get upgraded you'll need to somehow upgrade the Flatpak build files for all packages at once or things may not be able to work together anymore.

Yeah, I tend to agree with you here now. Looking into this issue, it's fairly clear that Flatpak was not designed with DAWs in mind. Perhaps Flatpak portals will one day allow for some layers of inter-process communication.

**For those who need a solution right now on a read-only destkop environment on a system like Fedora Silverblue**: I strongly recommend you look into [distrobox](https://github.com/89luca89/distrobox) in which you can create a virtual distribution instance (for example, Ubuntu 22.04) where you install bitwig and yabridge and `export` the bitwigstudio application. This will allow you to containerize your entire DAW working environment and isolate your dependencies from the rest of your system. This can also help stability, in theory, since your wine version / yabridge will only update when you explicitly enter the environment to do so. 

I might make a quick setup docker image that install yabridge and dependencies, with only bitwig needing to be supplied to the user.

**Comment by mitermayer** [comment]
Just hoping into this thread, I really would love to have a world where I can have a container with bitwig and vst support. Currently spending a considerable amount of time debugging why my wine stopped working

**Comment by korewaChino** [comment]
+1 for this. I'd really love to be able to install my plugins in a Bottles container and use my Windows DAWs inside the bottle, while also loading VST plugins from there.

**Comment by CleoMenezesJr** [comment]
I packaged Reaper to Flatpak, and now I'm just wondering the same thing. https://github.com/flathub/fm.reaper.Reaper
Any news? :/

**Comment by Eoin-ONeill-Yokai** [comment]
I've been looking into this a bit but I still don't know how to do it.

In theory, if you can create an extension for your DAW packaging that installs:
	- Wine version compatible with YB
	- Yabridge/YabridgeCTL

You could then run yabridgectl from the command line by doing something like (not verbatim, but in essence)

```
flatpak run --command=/bin/yabridgectl com.SomeCompany.SomeAwesomeDaw sync
```

In which case yabridge would be able to run w/ wine within the container, and yabridge would be able to export `.so` files appropriately. @CleoMenezesJr If you're in charge of managing the Reaper Flatpak you might understand all of this better than me, as most of my knowledge is only based on examining other flatpak repositories that deal with extensions (See Valve's Steam flatpak package with optional Gamescope Extension, which isn't flawless but is close.)

**Comment by rcsilva83** [comment]
I may have undestood wrong what you're discussing, but if you're talking about a Yabrigde Flatpak that would make any plugin available to a Flatpak DAW, I don't think it's the way it is supposed to be implemented. Audio plugins are available to Flatpak DAW's through plugins packaged in Flatpaks ([example](https://github.com/flathub/org.freedesktop.LinuxAudio.Plugins.x42Plugins/)). 

So, I think the way to use Yabridge in "Flatpak Way" would be to package a Windows plugin as a Flatpak, bundled with Wine and Yabridge.

**Comment by Eoin-ONeill-Yokai** [comment]
@rcsilva83 I think the problem with that implementation is that it assumes that people will be willing or have the rights to redistribute any plugin as a flatpak. For open source plugins, this is indeed a fine solution, but it's definitely not an option for plugins that are arbitrarily distributed (e.g. Kontact libraries, IK Multimedia, what have yous and what nots) since most of those cannot be redistributed without proper licensing.

**Comment by rcsilva83** [comment]
@Eoin-ONeill-Yokai yes, I agree with you. As I read your reply I had an idea but I don't know if it's feasible. If we had an Yabridge Flatpak, I think it would be easy enough to package most plugins in Flatpak. So easy that it could be done with a script. What if yabridgectl shipped such script? For example, it could be created a `yabridgectl flatpak` option that would build a Flatpak and install locally, so it could be used with Flatpak DAWs.

**Comment by abique** [comment]
One idea would be to have:
1. a clap plugin service daemon, it'd be reachable from flatpak or any app which has access to the daemon's socket
2. a proxy clap plugin, so the host speaks to the proxy clap plugin which itself speaks to the daemon to create a new plugin instance, and the to the plugin instance itself
3. a clap plugin host proxy: spawned by the daemon, and hosts the plugin

I recap:
- the daemon has knowledge of all the plugins avail on the system
- the daemon can create plugin instance and give you shm/socket to talk to it
- everything is transparent for both plugin and hosts, it is just a "proxy / man in the middle" technology
- this way we can distribute the plugin host in every flatpak runtimes, as well as for the system
- plugins can be sandboxed (no network, no file access, ...)

**Comment by abique** [comment]
The strong point of this approach is that it allows host and plugin built for different runtimes to work together.
In a way it is already what's yabridge is doing with wine: `bitwig <-> yabridge proxy <-> wine process with yabridge host <-> windows plugin`. Now that'd be: `bitwig <-> clap-daemon-proxy <-> clap-daemon` and once the plugin instance is created: `bitwig <-> clap-plugin-proxy <-> clap-host-proxy <-> plugin`.

Am I clear?

**Comment by rcsilva83** [comment]
@abique, I'm not familiar with CLAP. I just use Yabridge with a VST3 plugin. Is this deamon needed for VST3 too?

**Comment by dreamcat4** [comment]
> Am I clear?

If i am not misunderstanding... we were missing a clarification that [CLAP](https://en.wikipedia.org/wiki/CLever_Audio_Plug-in) is a new plugin standard to replace or go allongside VSTs.

So in this conversation it's a welcome 'also too'. Because of the possibility of porting existing JUCE library based VSTs over to CLAP. And because CLAP is just more modern, better, with fewer licencing issue too. Sine was alluded about being unable to distribute publically a VST flatpaks.

So it's most welcome. But allongside (both) since cannot replace (either-or) ?

**Comment by deusnovus** [comment]
For the sandbox thing to still be an issue with yabridge, I would imagine that Flatseal being able to provide filesystem access to every flatpak app is not that big of a development, no?

**Comment by robbert-vdh** [comment]
The situation is still the same. Someone would need to package both yabridge and Wine for the same Flatpak runtime the DAWs are running under. That wouldn't be too much effort, but it's something someone would need to do. You could also do it without packaging anything by just setting some oaths but that's not a very convenient or reproducible solution (and also, much more work than not simply not using Flatpak at all).

**Comment by ArtikusHG** [comment]
I have no idea why this hasn't been mentioned yet, but [Wine is already available as a Flatpak](https://flathub.org/apps/org.winehq.Wine) and has been for a while. It is unofficial, but looks like it's well and actively maintained - the last update was 3 months ago, and includes the current stable version of Wine released around the same time (8.0.2).

Packaging yabridge should (probably?) be significantly easier than packaging Wine. I might try to in my free time.

**Comment by xnghu** [comment]
+1 for wanting to see this happen, waiting patiently- while also reading up on material I need to learn to be able to help

**Comment by ejaa3** [comment]
Please read my mention, thanks.

**Comment by darltrash** [comment]
Any updates?

**Comment by photex** [comment]
Chiming in, Presonus Studio One Pro 7 is now in beta for Linux and is provided as flatpak for non-debian systems.

There are examples of a lot of common DAW plugins now available on flathub, and as expected there are multiple versions of each targeting different bases. In order for things to work you have to line that all up.

This is really annoying, and it really goes to show that flatpak and all the other new style linux application distribution methods like it (appimage, snap, ...) are not doing much other than perpetuating the existing problems with Linux rather than fixing anything.

But in either case, if there was some way to connect yabridge with a DAW flatpak (even if it was just a doc in the wiki explaining the gist of it all) that would be most welcome.

**Comment by jsonmaf1a** [comment]
+1 for this

**Comment by Milli189** [comment]
yeah I'm having an issue right now, where Bitwig flatpak detects the yabridge plugins, but says they have an error and so I'm unable to use them

**Comment by yioannides** [comment]
@Milli189 From the README:

<img width="897" height="247" alt="Image" src="https://github.com/user-attachments/assets/6f9dffee-098d-4bda-9b35-d98564633b97" />

**Comment by nullterminator-fyi** [comment]
Just a heads up for everyone, I have managed to get yabridge to work under flatpaked DAWs (3 as of now) and so I am currently working on an [experimental flatpak manifest for yabridge](https://github.com/nullterminator-fyi/yabridge-flatpak). But because of some limitations of `flatpak-builder` and the overall ecosystem, it might take me some time to get it to a usable state.

Until then, I will share my "hacks" that got yabridge to work under flatpak so more advanced users can use flatpaked DAWs in the meanwhile. This only involves setting some flatpak settings and installing some packages, so it shouldn't be too difficult to follow, albeit unfortunately it's gonna be very long. I will add additional notes for different yabridge and wine setups when necessary.

Here are the steps:
1. Make sure you have `yabridge` installed on your system. You may use your package manager, GitHub releases or build it yourself. Take note of the paths yabridge libraries and host exe's are installed. If you are unsure where did your package manager install them, you can run `which yabridge-host.exe` and `ldconfig -p | grep "libyabridge` respectively to find their paths (but most likely, they are located under `/usr/bin` and `/usr/lib`). Or use something called `host-os`, which I will explain later.
   - If you are using NixOS (like me), you are better off avoiding nix installations of yabridge for syncing plugins due to [this patch](https://github.com/NixOS/nixpkgs/blob/b79f5f98fc3e3c698012ed4ccedf9a6bc4fcb777/pkgs/tools/audio/yabridge/libyabridge-from-nix-profiles.patch).
2. Install wine to anywhere to your system. You can use your system package manager, extract a precompiled archive or use Bottles with [`yabridge-bottles-wineloader`](https://github.com/microfortnight/yabridge-bottles-wineloader). Bottles is recommended for easier wine version and prefix management. Also take note of the paths your wine binaries, prefixes and their libraries are located.
   - If you are using Bottles, make sure to install `yq` to your system as it's dependency of the bottles wineloader script. 
   - You either need to use a wine build with wow64 mode, or use a DAW with multiarch compatibility extensions and enabled `multiarch` setting; otherwise **_wine will not work and so your plugins_**. 
3. Open your flatpak permission settings. Installing Flatseal is the most distro-agnostic approach. If you use KDE, you can also find it under Settings > Applications > Flatpak Permission Settings.
4. Locate your DAW and add the necessary filesystem permissions: the user plugin paths (`~/.vst`, `~/vst3` and `~/.clap`), yabridge config path (`~/.config/yabridge`) and every wine prefix you use.
   - For Bottles, you can avoid allowing every single prefix path and just give your DAW access to Bottles' user data path (`.var/app/com.usebottles.bottles/data/bottles/`).
5. Give filesystem permissions for where your yabridge files are located. `~/.local/share/yabridge:ro` if you used github releases, the library and host.exe paths (with `:ro` appended for little security). if you installed with your package manager, It's likely going to be `/usr/lib:ro` and `/usr/bin:ro`.
   - Alternatively for package manager installations, you can also simply give access to `host-os:ro`
7. Give permissions for the paths of your wine binaries and libraries. If you installed with a system package manager, it's likely under the same parent paths as your yabridge libraries and host.exe's so you don't need to allow it again. If it's installed under a single directory, add it's path to the permissions.
   - If you use Bottles and already gave access to it's data path: make sure to give permission for the wineloader script's path (`~/.local/bin/`). Otherwise you can skip this step.
8. This one is complicated. Check if `LD_LIBRARY_PATH` and `PATH` are already defined. If they are, note their values somewhere. Let's call them `<previous value>`. Now for binary and library paths, if we are going to use them. Due to how flatpak works, `host-os` directories like `/usr`, `/bin` and `lib` are mounted under `/run/host/` in the sandbox. So for each of these paths, we will write `/run/host/<path>` replacing the `<path>` with the system path at our hand and we will call them <mounted path>. Rest of the paths will most likely stay the same. Knowing this, we will define `PATH` and `LD_LIBRARY_PATH` using these. If `PATH` is not defined, add `PATH=/app/bin:/usr/bin:<mounted path>:<yabridge host.exe path>:<wine binary path>`; If `LD_LIBRARY_PATH` is not defined, add `LD_LIBRARY_PATH=<mounted path>:<yabridge host.exe path>:<wine binary path>`. If `PATH` was already defined, set it to the value I described but replace `/app/bin:/usr/bin` with the previous value (like `PATH=<previous value>:<mounted path>:...`); If `LD_LIBRARY_PATH` was already defined, prepend them to the new value I described. Note that you cannot use `~` or other environment variables, since flatpak does not support expansion and substitution for env variables defined/overridden in manifests.
   - If you use Bottles with wineloader script, you don't need to add wine directories to neither `PATH` nor `LD_LIBRARY_PATH`, You just need to set `WINELOADER=/home/<user>/.local/bin/wineloader.sh`.

This all looks very confusing so I will give some case examples: 
   - Yabridge, wine, all installed with system package manager without predefined `PATH` or `LD_LIBRARY_PATH`: `PATH=/app/bin:/usr/bin:/run/host/usr/bin` & `LD_LIBRARY_PATH=/run/host/usr/lib`
   - Yabridge installed to `~/.local/share`, wine with Bottles and both ENVs not defined: `PATH=/app/bin:/usr/bin:/home/johndoe/.local/share/yabridge:/home/johndoe:/run/host/usr/bin` & `LD_LIBRARY_PATH=/home/johndoe/.local/share/yabridge:/run/host/usr/lib` & `WINELOADER=/home/johndoe/.local/bin/wineloader.sh` (we still added system bin/lib paths for `jq`.)
   - Yabridge installed with package manager, wine extracted from a precompiled archive to `~/.local/share/wine`, `PATH=/app/bin:/usr/bin:/app/plugins/bin` and `LD_LIBRARY_PATH=/app/plugins/lib` as initials: `PATH=/app/bin:/usr/bin:/app/plugins/bin:/run/host/usr/bin:/home/johndoe/.local/share/wine/bin` & `LD_LIBRARY_PATH=/app/plugins/lib:/run/host/usr/lib:/home/johndoe/.local/share/wine/lib`
9. Sync your plugins, and start scanning!

~(I will add one line flatpak commands here later on)~

I'm sorry that it became this long, but unfortunately it's this complicated. Majority of this will become unnecessary after I finish the flatpak manifest, but I still wanted to give early workarounds for advanced users, especially considering the rise of atomic distros. But at the end, it does end up working and I confirmed it on both my desktop PC and my laptop. Thought I didn't have enough time to test it all, so results may vary.

Here are some screenshots of these plugins running in different DAWs (do not worry, the sound is working):

<img width="1906" height="1071" alt="Image" src="https://github.com/user-attachments/assets/81994987-6998-4e5b-b085-22606163ab8f" />

Serum 2 filtered through supermassive in Bitwig

<img width="1906" height="1031" alt="Image" src="https://github.com/user-attachments/assets/eb5eed21-96fd-43d6-8299-722cc52ca7a1" />

Sforzando running in Ardour(this DAW was the most buggy of all)

<img width="1906" height="1031" alt="Image" src="https://github.com/user-attachments/assets/733f3bd7-ffa2-4b36-ae1f-1431c5f58f70" />

Tyrell.n6 in Bespoke

~Even if this guide looks intimidating, be patient and I might be able to make a yabridge flatpak a reality soon. In the meantime, I would greatly appreciate more testing so I can improve this guide. Have fun twiddling your knobs!~ I currently have a PoC manifest that somewhat works; but because of some difficulties and a current lack of passion, I haven't got much progress in the last few weeks and do not plan to write any documentation yet. I will get back on it once my finals end, so you can ask me anything by email until then. Thanks.
